Source/Noesis.Javascript/JavascriptFunction.cpp

```diff
@@ -0,0 +1,92 @@
+#include "JavascriptFunction.h"
+#include "JavascriptInterop.h"
+#include "JavascriptContext.h"
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+namespace Noesis { namespace Javascript {
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+JavascriptFunction::JavascriptFunction(v8::Handle<v8::Object> iFunction, JavascriptContext^ context)
+{
+	if (!iFunction->IsFunction())
+		throw gcnew System::ArgumentException("Trying to use non-function as function");
+	
+	if(!context)
+		throw gcnew System::ArgumentException("Must provide a JavascriptContext");
+
+	mFuncHandle = new Persistent<Function>(context->GetCurrentIsolate(), Handle<Function>::Cast(iFunction));
+	mContext = context;
+}
+
+JavascriptFunction::~JavascriptFunction()
+{
+	if(mFuncHandle) 
+	{
+		JavascriptScope scope(mContext);
+		mFuncHandle->Reset();
+		delete mFuncHandle;
+		mFuncHandle = nullptr;
+	}
+	System::GC::SuppressFinalize(this);
+}
+
+JavascriptFunction::!JavascriptFunction() 
+{
+	if(mFuncHandle) 
+	{
+		JavascriptScope scope(mContext);
```

Comment:
Regarding the access violation exception: It only occurs if the new `JavascriptFunction` object is used, either in the Fiddling project, or in the new unit tests - though there you only see the exception when debugging. We suspect that the GC calls the finalizer after the `JavascriptContext` has been destructed leading to a crash here. Is the `JavascriptScope` necessary? What does it do exactly? I just took it from the old branch.

Thanks!

Reply:
You should not be calling or otherwise accessing any v8 object after the context has been destroyed, because they will probably have been disposed by v8 and their memory freed.  I think you have these options:
* Don't do anything in the JavascriptFunction destructor, and allow mFuncHandle to be cleaned up when v8 is cleaned up.  (And document the memory leak somewhere.)
* Set a flag in JavascriptContext when it is disposed, which prevent JavascriptFunction from doing anything in its destructor when that flag is set.  This is obviously better because it won't leak memory, but is more complicated.